network/src/noise/handshake.rs

```diff
@@ -353,53 +347,44 @@ impl NoiseUpgrader {
-                        if !remote_pubkey_set.contains(&remote_public_key) {
-                            return Err(NoiseHandshakeError::UnauthenticatedClientPubkey(
-                                remote_peer_short,
-                                hex::encode(remote_public_key.as_slice()),
-                            ));
-                        }
-                    }
-                    None => {
-                        return Err(NoiseHandshakeError::UnauthenticatedClient(
-                            remote_peer_short,
-                            remote_peer_id,
-                        ))
-                    }
-                };
-                TrustLevel::Trusted
+                    Some(peer_auth_context) => Self::authenticate_inbound(
+                        remote_peer_short,
+                        peer_auth_context,
+                        &remote_public_key,
+                    ),
+                    None => Err(NoiseHandshakeError::UnauthenticatedClient(
+                        remote_peer_short,
+                        remote_peer_id,
+                    )),
+                }
             }
             HandshakeAuthMode::MaybeMutual(trusted_peers) => {
                 match trusted_peers.read().get(&remote_peer_id) {
-                    Some(remote_pubkey_set) => {
-                        if !remote_pubkey_set.contains(&remote_public_key) {
-                            return Err(NoiseHandshakeError::UnauthenticatedClientPubkey(
-                                remote_peer_short,
-                                hex::encode(remote_public_key.as_slice()),
-                            ));
-                        }
-                        TrustLevel::Trusted
-                    }
+                    Some(peer_auth_context) => Self::authenticate_inbound(
+                        remote_peer_short,
+                        peer_auth_context,
+                        &remote_public_key,
+                    ),
                     None => {
                         // if not, verify that their peerid is constructed correctly from their public key
                         let derived_remote_peer_id =
                             PeerId::from_identity_public_key(remote_public_key);
                         if derived_remote_peer_id != remote_peer_id {
-                            return Err(NoiseHandshakeError::ClientPeerIdMismatch(
+                            Err(NoiseHandshakeError::ClientPeerIdMismatch(
                                 remote_peer_short,
                                 remote_peer_id,
                                 derived_remote_peer_id,
-                            ));
+                            ))
+                        } else {
+                            Ok(PeerRole::Unknown)
                         }
-                        TrustLevel::Untrusted
                     }
                 }
             }
-        };
+        }?;
```

Comment:
why not just move this whole block into authenticate_inbound?

Reply:
Could do that, I basically just left it where it was

Reply:
So, here's why I did it this way.

This takes what's common between `mutual` and `maybe_mutual` auth paths and puts them in one place (so there's no copy differences).  Basically, the `Some` path is when we have a key to authenticate mutually, otherwise there's a divergence based on whether it allows unauthenticated peers